pytket/extensions/cutensornet/general_state/tensor_network_state.py

```diff
@@ -0,0 +1,466 @@
+
+        Returns:
+            Either a :code:`cupy.ndarray` on a GPU, or a :code:`numpy.ndarray` on a
+            host device (CPU). Arrays are returned in a 1D shape.
+        """
+        state_vector = cp.empty(
+            (2,) * self._circuit.n_qubits, dtype="complex128", order="F"
+        )
+        cutn.state_compute(
+            self._handle,
+            self._state,
+            self._work_desc,
+            (state_vector.data.ptr,),
+            self._stream.ptr,  # type: ignore
+        )
+        self._stream.synchronize()  # type: ignore
+        if on_host:
+            return cp.asnumpy(state_vector.flatten())
+        return state_vector.flatten()
+
+    def destroy(self) -> None:
+        """Destroys tensor network state."""
+        if self._work_desc is not None:
+            cutn.destroy_workspace_descriptor(self._work_desc)  # type: ignore
+        cutn.destroy_state(self._state)
+        del self._scratch_space
+
+
+class GeneralOperator:
+    """Handles tensor network operator."""
+
+    def __init__(
+        self,
+        operator: QubitPauliOperator,
+        num_qubits: int,
+        libhandle: CuTensorNetHandle,
+        loglevel: int = logging.INFO,
+    ) -> None:
+        """Constructs a tensor network operator.
+
+        From a list of Pauli strings and corresponding coefficients.
+
+        Args:
+            operator: The Pauli operator.
+            num_qubits: Number of qubits in a circuit for which operator is to be
+             defined.
+            libhandle: cuTensorNet handle.
+            loglevel: Internal logger output level.
+        """
+        # Mind the transpose for Y (same argument as in GeneralState)
+        self._pauli = {
+            "X": cp.asarray([[0, 1], [1, 0]]).astype("complex128", order="F"),
+            "Y": cp.asarray([[0, -1j], [1j, 0]]).T.astype("complex128", order="F"),
+            "Z": cp.asarray([[1, 0], [0, -1]]).astype("complex128", order="F"),
+            "I": cp.asarray([[1, 0], [0, 1]]).astype("complex128", order="F"),
+        }
+        self._logger = set_logger("GeneralOperator", loglevel)
+        self._handle = libhandle.handle
+        qubits_dims = (2,) * num_qubits
+        data_type = cq.cudaDataType.CUDA_C_64F  # TODO: implement a config class?
+        self._operator = cutn.create_network_operator(
+            self._handle, num_qubits, qubits_dims, data_type
+        )
+        self._logger.debug("Adding operator terms:")
+        for pauli_string, coeff in operator._dict.items():
+            if isinstance(coeff, Expr):
+                numeric_coeff = complex(coeff.evalf())  # type: ignore
+            else:
+                numeric_coeff = complex(coeff)  # type: ignore
+            self._logger.debug(f"   {numeric_coeff}, {pauli_string}")
+            num_pauli = len(pauli_string.map)
+            num_modes = (1,) * num_pauli
+            state_modes = tuple((qubit.index[0],) for qubit in pauli_string.map.keys())
```

Comment:
This is more problematic that it looked at first.

- As discussed above, this is incompatible with non-default qubit registers.
- Users cannot provide operators on a subset of the qubits in the circuit. The information of which qubits is selected is lost when it reaches `GeneralExpectationValue`. A workaround would be to add `I` operators, but `GeneralOperator` does not know which were the qubits in the original circuit, so it can't be done automatically for the user without altering the API.
- Looks like a nightmare to maintain, since `GeneralOperator` and `GeneralState` are using two different ways to address the qubits (as pointed out above), which is likely to cause subtle bugs in the future.

I don't think there's a point in adding a wrapper to `GeneralOperator`. We only want it for expectation values, so let's just manage it internally, transparently to the user. Similarly, I'm not convinced `GeneralExpectationValue` needs to be wrapped on its own object. I'll create a PR to this branch with a different approach where `GeneralState` is the only class which provides `statevector`, and `expectation_value` methods. This is just a high-level refactor of the code. With it, though, I should be able to solve the problems above, since the qubit addressing will be the same for operator and `GeneralState`.

Reply:
Additionally, the refactor I propose is more in line with the upcoming high-level API from NVIDIA, which would make an eventual transition easier.